File: ports/python/examples/langgraph_server.py
"""Hand-built LangGraph StateGraph on botiva — the Python counterpart of
examples/langgraph-server.ts, but deterministic (a rule-based agent node stands
in for the LLM, so no API key is needed and the selftest is CI-safe):

    START → agent ⇄ tools → END

It exercises every botiva context pattern in real LangGraph code:

  • the agent node reads UserStore via config["configurable"]["botiva"]
    (the explicit, portable pattern — identical in the TS/Go/.NET ports),
  • the remember_name tool writes UserStore via botiva_context() (ambient),
  • the get_weather tool pushes a GenUI card via botiva_emit(ui(...)) (ambient),
  • generate_report_pdf pauses with interrupt() → approval chips in the
    client; "Approve" resumes the tool via Command(resume=...).

Needs: pip install langgraph  (>= 0.2; no model provider required)

    cd ports/python
    python examples/langgraph_server.py               # server on :8796
    python examples/langgraph_server.py --selftest    # + scripted client, exit 0/1

Swap the rule-based agent node for a bound chat model (e.g. ChatAnthropic +
bind_tools) and everything else — transport, engine, HITL — stays identical.
"""

from __future__ import annotations

import asyncio
import json
import os
import re
import sys

from selftest_common import Client, bot_text, passed, tool_done, user_text  # noqa: E402 — also fixes sys.path

from botiva import ConversationEngine, botiva_context, botiva_emit, ui  # noqa: E402
from botiva.langgraph import LangGraphRuntime  # noqa: E402
from botiva.ws import WebSocketServer  # noqa: E402

try:
    from langchain_core.messages import AIMessage
    from langchain_core.runnables import RunnableConfig
    from langchain_core.tools import tool
    from langgraph.graph import END, START, MessagesState, StateGraph
    from langgraph.prebuilt import ToolNode
    from langgraph.types import interrupt

    try:
        from langgraph.checkpoint.memory import InMemorySaver
    except ImportError:  # older langgraph
        from langgraph.checkpoint.memory import MemorySaver as InMemorySaver
except ImportError:
    print(
        "This demo needs LangGraph:  pip install langgraph\n"
        "(the dependency-free demo is examples/server.py)",
        file=sys.stderr,
    )
    sys.exit(1)

PORT = int(os.environ.get("PORT", "8796"))

# ── tools — botiva_emit / botiva_context / interrupt() inside real tools ─────


@tool
async def get_weather(city: str) -> str:
    """Returns a city's weather and shows a weather card in the chat."""
    seed = sum(ord(ch) for ch in city)
    data = {
        "city": city,
        "temp": 12 + seed % 20,
        "condition": ["Sunny", "Partly Cloudy", "Rainy"][seed % 3],
        "humidity": 40 + seed % 40,
    }
    # Ambient emit — the framework knows which conversation/user this turn
    # belongs to; no plumbing through the graph.
    botiva_emit(ui("weather", data))
    return json.dumps(data)


@tool
async def remember_name(name: str) -> str:
    """Stores the user's name in their permanent profile (UserStore)."""
    ctx = botiva_context()
    if ctx is None:
        return "no ambient turn context"
    await ctx.user_store.patch({"name": name})
    return f"Saved. The user's name is {name} (persisted for user {ctx.user_id})."


@tool
def generate_report_pdf(topic: str) -> str:
    """Generates a report PDF on a topic (asks the user for approval first)."""
    # Human approval: the run pauses here (written to the checkpoint);
    # the user's next message returns via Command(resume=...).
    answer = interrupt({
        "question": f'Generate the "{topic}" report as PDF?',
        "options": ["Approve", "Cancel"],
    })
    if not re.search(r"approve|yes|onay|evet", str(answer), re.IGNORECASE):
        return "The user declined — no report was generated."
    file_name = "report-" + re.sub(r"[^a-z0-9]+", "-", topic.lower()) + ".pdf"
    botiva_emit(ui("genui-card", {
        "title": f"📄 {file_name}",
        "description": f'"{topic}" report is ready.',
        "actions": [{"label": "⬇️ Download", "value": f"download {file_name}"}],
    }))
    return f"Report ready: {file_name} (download card shown)."


TOOLS = [get_weather, remember_name, generate_report_pdf]

# ── the graph (hand-built; a rule-based "model" keeps it deterministic) ──────

_NAME_RE = re.compile(r"(?:my name is|ad[ıi]m)\s+(\w+)", re.IGNORECASE | re.UNICODE)
_ASK_NAME_RE = re.compile(r"what.*my name|ad[ıi]m\s+ne", re.IGNORECASE)
# City only when explicitly "weather in <city>"; a bare "weather"/"hava" falls
# through to the Istanbul default rather than grabbing the next word.
_WEATHER_RE = re.compile(r"weather\s+in\s+(\w+)|weather|hava", re.IGNORECASE | re.UNICODE)
_REPORT_RE = re.compile(r"report|pdf|rapor", re.IGNORECASE)

_call_seq = 0


def _tool_call(name: str, args: dict) -> AIMessage:
    global _call_seq
    _call_seq += 1
    return AIMessage(content="", tool_calls=[{"name": name, "args": args, "id": f"call-{_call_seq}"}])


async def agent_node(state: MessagesState, config: RunnableConfig) -> dict:
    """Rule-based stand-in for an LLM node. Reads UserStore through the
    portable config path — swap in `llm.bind_tools(TOOLS)` for the real thing."""
    messages = state["messages"]
    last = messages[-1]

    # A tool just ran → phrase the final answer from its result.
    if getattr(last, "type", "") == "tool":
        text = str(last.content)
        if getattr(last, "name", "") == "get_weather":
            text = "Here is the current weather."
        return {"messages": [AIMessage(content=text or "Done.")]}

    text = last.content if isinstance(last.content, str) else ""
    # Ask-name BEFORE name-set: "adım ne?" ("what's my name") otherwise matches
    # the name-set pattern and clobbers the stored name with "ne".
    if _ASK_NAME_RE.search(text):
        # Explicit TurnContext access — identical in TS/Go/.NET (PROTOCOL.md §9).
        botiva = (config.get("configurable") or {}).get("botiva")
        user = (await botiva.user_store.get() or {}) if botiva else {}
        reply = (
            f"Your name is {user['name']}."
            if user.get("name")
            else "I don't know your name yet — tell me with “my name is …”."
        )
        return {"messages": [AIMessage(content=reply)]}
    if m := _NAME_RE.search(text):
        return {"messages": [_tool_call("remember_name", {"name": m.group(1)})]}
    if (m := _WEATHER_RE.search(text)) is not None:
        return {"messages": [_tool_call("get_weather", {"city": m.group(1) or "Istanbul"})]}
    if _REPORT_RE.search(text):
        return {"messages": [_tool_call("generate_report_pdf", {"topic": "iteration velocity"})]}
    return {"messages": [AIMessage(content=f"Echo: {text}")]}


def route(state: MessagesState) -> str:
    last = state["messages"][-1]
    return "tools" if getattr(last, "tool_calls", None) else END


def build_graph():
    builder = StateGraph(MessagesState)
    builder.add_node("agent", agent_node)
    builder.add_node("tools", ToolNode(TOOLS))
    builder.add_edge(START, "agent")
    builder.add_conditional_edges("agent", route, ["tools", END])
    builder.add_edge("tools", "agent")
    return builder.compile(checkpointer=InMemorySaver())


# ── the server (same shape as every botiva server) ───────────────────────────


def build_engine() -> ConversationEngine:
    return ConversationEngine(
        LangGraphRuntime(build_graph()),
        greeting=(
            "Hi! LangGraph (Python) demo. Try: 'My name is Ada', "
            "'What's the weather in Istanbul?', or 'Generate a PDF report' 👋"
        ),
    )


async def selftest(url: str) -> None:
    a = await Client.connect(url)
    welcome = await a.wait_for(lambda f: f.get("type") == "welcome", "welcome")
    user_id = welcome["data"]["userId"]
    conversation_id = welcome["data"]["conversationId"]
    passed("welcome frame (protocol botiva/1)")

    await a.send("My name is Botivan, please remember it.")
    await a.wait_for(lambda f: tool_done(f, "remember_name"), "remember_name tool")
    await a.wait_for(lambda f: bot_text(f, "Botivan"), "name confirmation")
    passed("remember_name tool → UserStore write (ambient botiva_context)")

    await a.send("What's the weather in Istanbul?")
    await a.wait_for(
        lambda f: f.get("type") == "genui" and (f.get("chunk") or {}).get("component") == "weather",
        "weather genui card",
    )
    passed("get_weather → botiva_emit GenUI card")

    await a.send("Generate a PDF report about velocity")
    await a.wait_for(lambda f: f.get("type") == "text" and f.get("actions"), "interrupt chips")
    passed("interrupt() → approval chips")

    await a.send("Approve")
    await a.wait_for(lambda f: tool_done(f, "generate_report_pdf"), "resume completes tool")
    await a.wait_for(
        lambda f: f.get("type") == "genui" and (f.get("chunk") or {}).get("component") == "genui-card",
        "download card after resume",
    )
    await a.wait_for(lambda f: bot_text(f, "Report ready"), "final answer after resume")
    passed("Command(resume) → tool completed + download card")

    b = await Client.connect(f"{url}?userId={user_id}&conversationId={conversation_id}&watermark=0")
    await b.wait_for(lambda f: user_text(f, "My name is Botivan"), "replay")
    passed("watermark replay on reconnect")

    c = await Client.connect(f"{url}?userId={user_id}")
    await c.wait_for(lambda f: f.get("type") == "welcome", "welcome (C)")
    await c.send("What is my name?")
    await c.wait_for(lambda f: bot_text(f, "Botivan"), "name recalled across conversations")
    passed("new conversation, same userId → UserStore recalled in the agent node")

    await a.close()
    await b.close()
    await c.close()


async def main() -> int:
    server = WebSocketServer(build_engine(), port=PORT)
    await server.start()
    print(f"\n✓ botiva LangGraph (Python) demo ready → ws://localhost:{PORT}/chat\n")

    if "--selftest" in sys.argv:
        try:
            await selftest(f"ws://localhost:{PORT}/chat")
        except (TimeoutError, AssertionError, ConnectionError) as err:
            print(f"\nLangGraph selftest failed ❌ {err}", file=sys.stderr)
            return 1
        finally:
            await server.close()
        print("\nLangGraph selftest passed ✅")
        return 0

    await server.serve_forever()
    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: examples/ts/langgraph-server.ts
// Hand-built LangGraph StateGraph + Claude on botiva — the LLM-powered version
// of the pattern this framework was designed around. Unlike demo-server.ts
// (which uses the prebuilt createReactAgent), the graph here is wired by hand:
//
//   START → agent ⇄ tools → END          (agent = ChatAnthropic with bound tools)
//
// and shows every botiva context pattern in real agent code:
//
//   • agent node reads UserStore via config.configurable.botiva (the explicit,
//     portable pattern — identical in the Go/.NET/Python ports) and injects
//     what it knows about the user into the system prompt,
//   • remember_name tool writes UserStore via botivaContext() (ambient),
//   • get_weather tool pushes a GenUI card via botivaEmit(ui(...)) (ambient),
//   • generate_report_pdf pauses with interrupt() → approval chips in the
//     client; "Approve" resumes the tool via Command({ resume }).
//
//   pnpm demo:graph                       # server on :8792 (ws + socket.io)
//   pnpm exec tsx examples/langgraph-server.ts --selftest   # live LLM check
//
// Needs ANTHROPIC_API_KEY (or anthropic-key.txt in the parent repo root).
// For the deterministic, LLM-free end-to-end test use `pnpm smoke` instead.

import { createServer } from "node:http";
import express from "express";
import { Server as SocketIOServer } from "socket.io";
import { ChatAnthropic } from "@langchain/anthropic";
import { SystemMessage, type AIMessage } from "@langchain/core/messages";
import { tool } from "@langchain/core/tools";
import { END, interrupt, MemorySaver, MessagesAnnotation, START, StateGraph } from "@langchain/langgraph";
import { ToolNode } from "@langchain/langgraph/prebuilt";
import { z } from "zod";
import {
    botivaContext,
    botivaEmit,
    ConversationEngine,
    ui,
    type TurnContext,
} from "@botiva/core";
import { WebSocketConnector } from "@botiva/websocket";
import { SocketIOConnector } from "@botiva/socket.io";
import { LangGraphRuntime } from "@botiva/langgraph";
// @ts-ignore parent repo helper (plain JS)
import { loadAnthropicKey } from "../../../langgraph/config.mjs";

const port = Number(process.env.PORT ?? 8792);
loadAnthropicKey();

// ── tools — botivaEmit / botivaContext / interrupt() inside real tools ──────

const getWeather = tool(
    async ({ city }: { city: string }) => {
        const seed = [...city].reduce((acc, ch) => acc + ch.charCodeAt(0), 0);
        const data = {
            city,
            temp: 12 + (seed % 20),
            condition: ["Sunny", "Partly Cloudy", "Rainy"][seed % 3],
            humidity: 40 + (seed % 40),
        };
        // Ambient emit — the framework knows which conversation/user this
        // turn belongs to; no plumbing through the graph.
        botivaEmit(ui("weather", data));
        return JSON.stringify(data);
    },
    {
        name: "get_weather",
        description: "Returns a city's weather and shows a weather card in the chat.",
        schema: z.object({ city: z.string().describe("e.g. Istanbul") }),
    },
);

const rememberName = tool(
    async ({ name }: { name: string }) => {
        // Ambient TurnContext inside a tool: write UserStore — this survives
        // across conversations and devices for the same userId.
        const ctx = botivaContext();
        await ctx?.userStore.patch({ name });
        return `Saved. The user's name is ${name} (persisted for user ${ctx?.userId ?? "?"}).`;
    },
    {
        name: "remember_name",
        description: "Stores the user's name in their permanent profile (UserStore).",
        schema: z.object({ name: z.string() }),
    },
);

const generateReportPdf = tool(
    async ({ topic }: { topic: string }) => {
        // Human approval: the run pauses here (written to the checkpoint);
        // the user's next message returns via Command({ resume }).
        const answer = interrupt({
            question: `Generate the "${topic}" report as PDF?`,
            options: ["Approve", "Cancel"],
        });
        if (!/approve|yes|onay|evet/i.test(String(answer))) {
            return "The user declined — no report was generated.";
        }
        const fileName = `report-${topic.toLowerCase().replace(/[^a-z0-9]+/g, "-")}.pdf`;
        botivaEmit(
            ui("genui-card", {
                title: `📄 ${fileName}`,
                description: `"${topic}" report is ready.`,
                actions: [{ label: "⬇️ Download", value: `download ${fileName}` }],
            }),
        );
        return `Report ready: ${fileName} (download card shown).`;
    },
    {
        name: "generate_report_pdf",
        description: "Generates a report PDF on a topic (asks the user for approval first).",
        schema: z.object({ topic: z.string() }),
    },
);

const tools = [getWeather, rememberName, generateReportPdf];

// ── the graph (hand-built — no createReactAgent) ────────────────────────────

const llm = new ChatAnthropic({
    model: process.env.CLAUDE_MODEL ?? "claude-opus-4-8",
    maxTokens: 1500,
}).bindTools(tools);

type GraphState = typeof MessagesAnnotation.State;

/** Agent node — reads UserStore through the portable config path. */
async function agentNode(state: GraphState, config?: { configurable?: { botiva?: TurnContext } }) {
    const botiva = config?.configurable?.botiva;
    const user = ((await botiva?.userStore.get()) ?? {}) as { name?: string };
    const system = new SystemMessage(
        [
            "You are the botiva LangGraph demo assistant.",
            "When the user tells you their name, ALWAYS call remember_name.",
            "For weather questions ALWAYS call get_weather; the card is shown automatically, so answer in one short sentence without repeating the numbers.",
            "If the user asks for a PDF/report, call generate_report_pdf directly WITHOUT asking for confirmation — the tool asks for approval itself.",
            user.name
                ? `You already know this user: their name is ${user.name} (from UserStore, it survives across conversations).`
                : "",
            "Reply in the user's language, short and to the point.",
        ]
            .filter(Boolean)
            .join(" "),
    );
    const response = await llm.invoke([system, ...state.messages]);
    return { messages: [response] };
}

const route = (state: GraphState) => {
    const last = state.messages.at(-1) as AIMessage | undefined;
    return last?.tool_calls?.length ? "tools" : END;
};

const graph = new StateGraph(MessagesAnnotation)
    .addNode("agent", agentNode)
    .addNode("tools", new ToolNode(tools))
    .addEdge(START, "agent")
    .addConditionalEdges("agent", route, ["tools", END])
    .addEdge("tools", "agent")
    .compile({ checkpointer: new MemorySaver() });

// ── the server (same shape as every botiva server) ──────────────────────────

const engine = new ConversationEngine({
    runtime: new LangGraphRuntime(graph),
    greeting:
        "Hi! Hand-built LangGraph + Claude demo. Try: 'My name is Ada', 'What's the weather in Istanbul?', or 'Generate a PDF report about street food' 👋",
});

const app = express();
app.get("/healthz", (_req, res) => {
    res.json({ ok: true, engine: "botiva-langgraph-llm-demo" });
});
const server = createServer(app);
new WebSocketConnector({ engine, server });
const io = new SocketIOServer(server, { cors: { origin: "*" } });
new SocketIOConnector({ engine, io });

await new Promise<void>((resolve) => server.listen(port, resolve));
console.log(`\n✓ botiva LangGraph demo (LLM) ready → ws://localhost:${port}/chat (and socket.io)\n`);

// ── optional scripted self-test (live LLM — needs the API key) ──────────────

if (process.argv.includes("--selftest")) {
    const { default: WebSocket } = await import("ws");
    const LLM_TIMEOUT = 90_000;

    class Client {
        frames: any[] = [];
        #listeners = new Set<(f: any) => void>();
        #ws: InstanceType<typeof WebSocket>;

        constructor(query = "") {
            this.#ws = new WebSocket(`ws://localhost:${port}/chat${query}`);
            this.#ws.on("message", (raw) => {
                const frame = JSON.parse(raw.toString());
                this.frames.push(frame);
                for (const listener of [...this.#listeners]) listener(frame);
            });
        }

        open() {
            return new Promise((resolve, reject) => {
                this.#ws.on("open", resolve);
                this.#ws.on("error", reject);
            });
        }

        send(text: string) {
            this.#ws.send(JSON.stringify({ type: "text", data: { text } }));
        }

        close() {
            this.#ws.close();
        }

        /** Wait for a frame matching pred, arriving at index >= from. */
        waitFor(pred: (f: any) => boolean, label: string, from = 0): Promise<any> {
            const existing = this.frames.slice(from).find(pred);
            if (existing) return Promise.resolve(existing);
            return new Promise((resolve, reject) => {
                const timer = setTimeout(() => {
                    this.#listeners.delete(check);
                    reject(new Error(`timeout: ${label}`));
                }, LLM_TIMEOUT);
                const check = (f: any) => {
                    if (!pred(f)) return;
                    clearTimeout(timer);
                    this.#listeners.delete(check);
                    resolve(f);
                };
                this.#listeners.add(check);
            });
        }

        /** Send a message, then await pred AND the end of that run. */
        async turn(text: string, pred: (f: any) => boolean, label: string) {
            const from = this.frames.length;
            this.send(text);
            const frame = await this.waitFor(pred, label, from);
            await this.waitFor(
                (f) => f.type === "run" && f.data?.status === "finished",
                `${label} (run finished)`,
                from,
            );
            return frame;
        }
    }

    const toolDone = (f: any, name: string) =>
        f.type === "tool_call" && f.data?.name === name && f.data?.status === "completed";
    const botText = (f: any, re: RegExp) =>
        f.type === "text" && f.from === "bot" && re.test(f.data?.text ?? "");

    try {
        const a = new Client();
        await a.open();
        const welcome = await a.waitFor((f) => f.type === "welcome", "welcome");
        const userId = welcome.data.userId;

        await a.turn(
            "My name is Botivan, please remember it.",
            (f) => toolDone(f, "remember_name"),
            "remember_name tool via LLM",
        );
        console.log("  ✅ LLM called remember_name → UserStore write");

        const weather = await a.turn(
            "What's the weather in Istanbul?",
            (f) => f.type === "genui" && f.chunk?.component === "weather",
            "weather genui via botivaEmit",
        );
        console.log(`  ✅ get_weather → botivaEmit GenUI card (streamId ${weather.streamId ? "ok" : "MISSING"})`);

        await a.turn(
            "Generate a PDF report about street food.",
            (f) => f.type === "text" && Array.isArray(f.actions) && f.actions.length > 0,
            "interrupt approval chips",
        );
        console.log("  ✅ generate_report_pdf paused via interrupt() → approval chips");

        const fromApprove = a.frames.length;
        a.send("Approve");
        await a.waitFor((f) => toolDone(f, "generate_report_pdf"), "resume completes tool", fromApprove);
        await a.waitFor(
            (f) => f.type === "genui" && f.chunk?.component === "genui-card",
            "download card after resume",
            fromApprove,
        );
        console.log("  ✅ Command({resume}) → tool completed + download card");

        // New conversation, same user → the agent node reads UserStore.
        const b = new Client(`?userId=${userId}`);
        await b.open();
        await b.waitFor((f) => f.type === "welcome", "welcome (B)");
        await b.turn("What is my name?", (f) => botText(f, /Botivan/i), "name recalled across conversations");
        console.log("  ✅ new conversation, same userId → agent recalled the name from UserStore");

        a.close();
        b.close();
        console.log("\nLangGraph LLM selftest passed ✅");
        process.exit(0);
    } catch (err) {
        console.error(`\nLangGraph LLM selftest failed ❌ ${(err as Error).message}`);
        process.exit(1);
    }
}
